core/src/main/java/org/testcontainers/containers/GenericContainer.java

```diff
@@ -300,36 +329,73 @@ private void tryStart() {
             CreateContainerCmd createCommand = dockerClient.createContainerCmd(dockerImageName);
             applyConfiguration(createCommand);
 
-            containerId = createCommand.exec().getId();
+            createCommand.getLabels().put(DockerClientFactory.TESTCONTAINERS_LABEL, "true");
 
-            connectToPortForwardingNetwork(createCommand.getNetworkMode());
+            if (shouldBeReused) {
+                if (!canBeReused()) {
+                    throw new IllegalStateException("This container does not support reuse");
+                }
+
+                if (TestcontainersConfiguration.getInstance().environmentSupportsReuse()) {
+                    String hash = hash(createCommand);
+
+                    containerId = findContainerForReuse(hash).orElse(null);
```

Comment:
the initial implementation was inlined, but then I extracted a method to make @kiview happy =P